rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/RMNodeStarter.java

```diff
@@ -432,6 +435,63 @@ public static void main(String[] args) {
         }
     }
 
+    private static BlockingQueue<KeyPair> keyPairs = new LinkedBlockingQueue<>();
+
+    private static void launchBackgorundKeyPairMaker(final int numberOfWorkers) {
+        final Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (!Thread.interrupted()) {
+                    try {
+                        // if we obeserve less than 'numberOfWorkers' elements than we goint to
+                        // add 'numberOfWorkers' more
+                        if (keyPairs.size() < numberOfWorkers) {
+                            RMNodeStarter.generateAndAddKeyPairs(numberOfWorkers);
+                        }
+
+                        // sleep 1 second, ot until notification (from getKeyPair method)
+                        synchronized (keyPairs) {
+                            keyPairs.wait(1000);
+                        }
+                    } catch (InterruptedException | NoSuchAlgorithmException e) {
+                        logger.warn("Something went terribly wrong.", e);
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            }
+        });
+        thread.start();
+    }
+
+    private static void generateAndAddKeyPairs(int numberOfWorkers)
+            throws NoSuchAlgorithmException, InterruptedException {
+        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
+        keyGen.initialize(1024, new SecureRandom());
+        KeyPair keyPair = keyGen.generateKeyPair();
+        keyPairs.put(keyPair);
```

Comment:
I don't see the numberOfWorkers parameter being used, but your algorithm admits that several keypairs can be generated at each cycle. Also, use offer instead of put (equivalent, but more associated with queues)

Reply:
True, tremendous bug. Thanks